hoomd/mpcd/update.py

```diff
@@ -0,0 +1,132 @@
+# Copyright (c) 2009-2024 The Regents of the University of Michigan.
+# Part of HOOMD-blue, released under the BSD 3-Clause License.
+
+r"""MPCD reverse perturbation to establish nonequilibrium shear flow.
+
+These are MPCD updaters.
+
+.. invisible-code-block: python
+
+    simulation = hoomd.util.make_example_simulation(mpcd_types=["A"])
+    simulation.operations.integrator = hoomd.mpcd.Integrator(dt=0.1)
+
+"""
+
+from . import _mpcd
+from hoomd.operation import Updater
+from hoomd.data.parameterdicts import ParameterDict
+from hoomd.logging import log
+from hoomd.data.typeconverter import OnlyTypes, positive_real
+
+import math
+
+
+class ReverseNonequilibriumShearFlow(Updater):
+    r"""MPCD reverse nonequilibrium shear flow.
+
+    Args:
+        trigger (hoomd.trigger.trigger_like): Trigger to swap momentum.
+
+        num_swaps (int): Maximum number of particle pairs to swap.
+
+        slab_width (float): Thickness of momentum-exchange slabs.
+
+        target_momentum (float): Target momentum for swapped particles.
+
+    There are two exchange slabs separated by a distance of :math:`L_y/2`. The
+    edges of these slabs are located at (:math:`-L_y/2`, :math:`-L_y/2` +
+    `slab_width`) and (:math:`0.0`, `slab_width`) along the *y*-direction
+    (gradient direction) of the simulation box. On each `trigger`, particles are
+    sorted into the exchange slabs based on their positions in the box. In each
+    slab, particles whose *x*-component momenta are near `target_momentum` are
+    selected for a pairwise momentum swap, and up to `num_swaps` swaps are
+    executed. Candidate particles for swapping are those with positive
+    *x*-component momentum (flow direction) in the lower slab and those with
+    negative *x*-component momentum in the upper slab.
+
+    To simulate the original implementation by
+    `Müller-Plathe <https://doi.org/10.1103/PhysRevE.59.4894>`_,
+    only the fastest particle and the slowest particle are swapped. This is
+    achieved by setting `num_swaps` to 1 and `target_momentum` to a
+    sufficiently large value. It is also feasible for particles that are not
+    necessarily the fastest nor slowest to be selected for momentum exchange
+    as `Tenney and Maginn <https://doi.org/10.1063/1.3276454>`_ did. This is
+    accomplished by setting the `target_momentum` argument to a smaller value.
+
+    This swapping of particle momenta imposes a momentum flux on the system,
+    which drives the flow.
+
+    .. rubric:: Example:
+
+    .. code-block:: python
+
+            flow = hoomd.mpcd.update.ReverseNonequilibriumShearFlow(
+                trigger=20, num_swaps=10, slab_width=0.25, target_momentum=5.0)
+            simulation.operations.updaters.append(flow)
```

Comment:
The example should first show how to implement the standard Muller-Plathe method. Then, you can add another example of the other method. I would actually suggest that you move those short text blurbs down from the paragraph above and put them before each of the two example code blocks so there is a code example going with the description.
```suggestion
    .. code-block:: python

            flow = hoomd.mpcd.update.ReverseNonequilibriumShearFlow(
                trigger=1, num_swaps=1, slab_width=1)
            simulation.operations.updaters.append(flow)
```